### DISCUSSION TITLE
Realtime API Websockets

### DISCUSSION THREAD

**Discussion by Fusseldieb** [directus/directus#3255] [Feature Requests]
**As** proposed by @rijkvanzanten in #3221 

> feature request for `ws://` (and `wss://`) support for the REST endpoints

> ideally somewhat of a spec what the IO should look like? ✨

Like I said, I don't feel qualified to lay down the IO specs, but here we go/try anyways:

#### 📝 Subscribing to a collection

- To receive updates, the Client/App should probably "subscribe" itself to the current collection or item, so that it doesn't receive "spam" from irrelevant collections that aren't within the users view. This could happen as a JSON request like `{ "action": "subscribe", "collection": "test", "interval": "5000", "data": { <same data as HTTP API (GET)> } }`.

Where:
- `action` could be `subscribe`, `unsubscribe` or any of the REST methods like `get`,`post`, etc. When subscribing for the first time in a particular collection, the websocket should probably also send the entire collection along the way, thus the `data` key, so it could subscribe as soon as the page loads and already receive the entire collection to display, without needing a second request. This would improve latency, especially on further away servers with a higher ping.
- `collection` The current open collecion which we want to receive updates from. *
- `interval` the interval *in milliseconds* the API/WS should send us updates about the subscribed collection or item. 
While this *could* be done in the frontend, on heavily active projects it would waste a lot of bandwidth and also performance/speed 
since the websocket would be spamming changes to the client, while the client ignores most of it.
- _* Note that the parameters for the `GET` request (in general) would be put directly inside the root Object, like above, not inside `data`. `data` would contain the same stuff as the REST HTTP API. With that, it would be possible to write the docs for both the API/WS and API/HTTP at once, saving time and possibly confusion._
- _The key `collection` would tell us what we want (possible: `collection`, `auth`, ...), similar to the *path* in the HTTP API. This, among many other things, might be worth discussing first._

#### ❌ Unsubscribing from a collection

Similar to `subscribe`: `{ "action": "unsubscribe", "collection": "test" }`

#### ❤ Heartbeat / Token Refresh

In order to track if the connection is still alive and to refresh possible tokens or whatever, a simple message should be sent back and forth every so often (10 seconds or so?). 

Since we already have a HTTP API endpoint which does just this, we could reuse it and do it like: `{ "auth": "refresh" }` or simply slightly taking a new road for this and do: `{ "action": "ping" }` which replies with a pong or similar.

#### 📃 `POST`ing an Item

 To submit an item, it could be sent similar to: `{ "action": "post", "collection": "test", "data": { <same data as HTTP API> } }`
Followed by the same response as in the HTTP API docs or similar.

#### 👀 `GET`ing an Item

 To get an item, it could be sent similar to: `{ "action": "get", "collection": "test", "id": 12, "data": { <same data as HTTP API> } }`.

And so on...

The responses could be equal or similar to those which we would receive when doing a HTTP API request.

I hope I'm not missing something crucial :)

Would love to hear thoughts on this!

**Comment by rijkvanzanten** [comment]
I'm wondering if we should have a single `ws://` endpoint with ways to subscribe to certain collections, or make "every" endpoint `ws` compatible 🤔 

A solution for the subscribe-to-only-a-single-collection would be to only hit up `ws://example.com/items/articles`

**Reply by rijkvanzanten** [reply]
> To submit an item, it could be sent similar to: { "action": "post", "collection": "test", "data": { <same data as HTTP API> } }
Followed by the same response as in the HTTP API docs or similar.

I love the idea that the payload is effectively a way to describe a REST request in JSON form 👍

**Comment by emregency** [comment]
> #### Subscribing to a collection
> * To receive updates, the Client/App should probably "subscribe" itself to the current collection or item, so that it doesn't receive "spam" from irrelevant collections that aren't within the users view. This could happen as a JSON request like `{ "action": "subscribe", "collection": "test", "interval": "5000", "data": { <same data as HTTP API (GET)> } }`.

I really like @Fusseldieb 's approach of initiating a subscription to collections through a JSON request.

> * interval the interval in milliseconds the API/WS should send us updates about the subscribed collection or item.
While this could be done in the frontend, on heavily active projects it would waste a lot of bandwidth and also performance/speed
since the websocket would be spamming changes to the client, while the client ignores most of it.

Isn't this polling @Fusseldieb? I would leave this out of the ws/subscription. You can always poll at different intervals from anywhere. Maybe I am missing something here.

**Reply by rijkvanzanten** [reply]
I think they describe it more as a throttle on the amount of messages sent from the WS connection so it only sends messages every X ms instead of live to save some overhead 🤔

**Reply by Fusseldieb** [reply]
> Isn't this polling @Fusseldieb? I would leave this out of the ws/subscription. You can always poll at different intervals from anywhere. Maybe I am missing something here.

Actually, no. I'm not polling the API. The API would send me updates every few seconds (or mins, depending) about changes in the API on its own after being subscribed to a collection. 
Theoretical benefits could include:
- Less use of bandwidth 
_Since it doesn't need two messages to answer with the changes aka `"send me the changes" -> "here are the changes"`_
- Reduced delay when something changes
_The API would push the changes immediately to the client which in turn would be updating the UI asap, while on the other hand polling it from the client side would only update things if asked to the server first, which would introduce delay._

> I think they describe it more as a throttle on the amount of messages sent from the WS connection so it only sends messages every X ms instead of live to save some overhead 🤔

Yes, I meant basically that. Setting the interval up would mean less messages over WS, but more data in the actual packet when it finally arrives (more changes would pile up)

**Reply by emregency** [reply]
All good then! 😅

**Comment by pwFoo** [comment]
Websocket works with active / opened clients, but as far as I know not for example with service-workers? Service-worker close connections after some (short) time. So websockets could help to keep the app data up to date without polling, but only for active used apps. 
Push notifications with webpush api works with closed apps (for example closed browser / tab, but receive notifications from android os or browser background job) and could trigger a service-worker to do some update jobs. 

Push notitification handling (save registrations, use for webhooks / defined actions on changes) would be much more power for a directus app?

**Reply by rijkvanzanten** [reply]
Sure, but it sounds like that's a different feature request / discussion unrelated to websockets support. Different feature for a different goal 🙂

**Reply by pwFoo** [reply]
Hi... 
Thats right... I think push would be more important, but websockets are great too :)

Polling would fit many cases for applications. Websockets (or EventSource based streams) as optional feature without performance issues for the rest api would be nice.

**Comment by u12206050** [comment]
Just a word of caution with websockets. We have a product that was initially built with featherjs using websockets for real time updates. The initial development and testing was fun and cool, but once it came to scaling it became very difficult. Using GCP with open websockets connections took some time to get working properly. Eventually we switched back to normal http requests, however pushing data updates to firestore and having clients subscribe to collections in firestore.

Obviously permissions had to be synced and a firestore token generated per user with the claims they have for the collections we wanted real time updates for.

So in summary. Beware of building WS support into a platform that needs to be able to scale horizontally.

Open for anyone to show me working scalable ws solution without a native hosted service eg. PubSub, firstore, pusher etc.

**Reply by rijkvanzanten** [reply]
Horizontally scaling WS is definitely tricky, and would require a shared pubsub backend to function. We'll most likely require the use of Redis for multi-instance horizontally scaled directus installations w/ websockets

**Reply by drnachio** [reply]
Redis will be your best friend scaling WS

**Reply by cliqer** [reply]
Check KeyDB (a Redis alternative with Master-to-Master replication) and Cloudflare Workers with KV for a global signalling/database system.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
This is merged and launched in version 10.3! 🚀

**Reply by Fusseldieb** [reply]
This is awesome to hear :)
